net/sslh/src/etc/inc/plugins.inc.d/sslh.inc

```diff
@@ -0,0 +1,74 @@
+
+    1. Redistributions of source code must retain the above copyright notice,
+       this list of conditions and the following disclaimer.
+
+    2. Redistributions in binary form must reproduce the above copyright
+       notice, this list of conditions and the following disclaimer in the
+       documentation and/or other materials provided with the distribution.
+
+    THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
+    INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
+    AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+    AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
+    OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+
+/**
+ * Function to add plugin to the services page of OPNsense.
+ *
+ * This adds the service to the System: Diagnostics: Services page at
+ * http://<opnsense>/status_services.php
+ *
+ * The suffix "_services" allows this function to be picked up by
+ * etc/inc/plugins.inc:plugins_services() and be included in the
+ * $services array to be processed by status_services.php.
+ *
+ * @return array the array of attributes for this service
+*/
+function sslh_services()
+{
+
+    // Create an array to be processed by www/status_services.php
+    $service = array();
+    // Load the plugin configuraiton file as an array for use below.
+    $conf = parse_ini_file('/usr/local/opnsense/service/conf/sslh.conf', true);
+    // Load in our settings to get the enabled state of the plugin.
+    $settings = new \OPNsense\Sslh\Settings();
+
+    // Only show the plugin if it's enabled.
+    if (! ((string) $settings->enabled == '1')) {
+        // return empty array if not enabled
+        return $services;
+    }
+
+    $configd_name = $conf['plugin']['configd_name'];
+    $service[] = array(
+        'name' => $conf['plugin']['name'],                  // Service column
+        'description' => gettext($conf['plugin']['label']), // Description column
+        'configd' => array(                                 // Status column
+            'restart' => array($configd_name . ' restart'),
+            'start' => array($configd_name . ' start'),
+            'stop' => array($configd_name . ' stop'),
+        ),
+        // 'pidfile' is used by etc/inc/util.inc:service_status(),
+        // and displays service state in status column. If undefined,
+        // the 'name' as defined above is used in is_process_running()
+        // which must match the process name of the service in order to be found.
+        'pidfile' => $conf['plugin']['pidfile'],
```

Comment:
All things considered the question is if it's worth it. The pid file is usually also defined in the rc.d file of the service that comes with the port. Softcoding it in a second place seems less favourable than hardcoding it once here.

Reply:
I happened across this by accident since I copied this file from the dnscrypt-proxy plugin which uses `pid` as the key name for this value.
https://github.com/opnsense/plugins/blob/master/dns/dnscrypt-proxy/src/etc/inc/plugins.inc.d/dnscryptproxy.inc#L51

Since for dnscrypt-proxy, it happens that the service name and the process name actually match, everything works without issue. However, with sslh, the process names are one of either `sslh-fork` or `sslh-select` which doesn't match the 'name' value in the array, and thus it's not able to find the process by name. It only shows the service as stopped in the status column. So, for this case, the `pidfile` value is how `is_process_running()` is able to find the process and properly detect if it's running or not.

For this case, the pid file is a command line option/configuration option, and is explicitly defined in the configuration file. It's also defined in the conf for reference here. I also looked in the `/usr/local/etc/rc.d/sslh` but for sslh it didn't look like one was defined as I've seen in other services.

Reply:
It seems that is a bug worth fixing?

```
# git grep "'pid'"
dns/dnscrypt-proxy/src/etc/inc/plugins.inc.d/dnscryptproxy.inc:        'pid' => '/var/run/dnscrypt-proxy.pid'
net-mgmt/nrpe/src/etc/inc/plugins.inc.d/nrpe.inc:        'pid' => '/var/run/nrpe.pid'
net/ntopng/src/etc/inc/plugins.inc.d/ntopng.inc:        'pid' => '/var/run/ntopng/ntopng.pid'
```

Well, the 'name' value should match the process name (binary), at least that what is being used if no PID file is given. In the example you give it seems perfectly fine except for the key not being proper. Especially for plugins.inc/d content the file is always bound to be adapted by plugin maintainer.

Reply:
I'll take a look at the other plugins and submit PRs for those as well.